keps/sig-api-machinery/storage-migration.md

```diff
@@ -0,0 +1,346 @@
+  the completed `job`, the `configmap` is garbage collected.
+
+### Beta workflow - Automation
+
+It is a beta goal to automate the migration workflow. That is, migration does
+not need to be triggered manually by cluster admins, or by custom control loops
+of Kubernetes distributions.
+
+The automated migration should work for Kubernetes built-in resource types,
+custom resources, and aggregated resources.
+
+Here are two alternatives and their requirements.
+
+#### a. Exposing the configured storage versions via API
+
+This is what the [Active Management][] design proposes. It requires:
+* exposing the configured [storage versions][] of Kubernetes built-in resource
+  types and aggregated resources. 
+
+Note that custom resources already exposes this information. 
+
+The migration then is triggered by updates of storage versions.
+
+[Active Management]:https://docs.google.com/document/d/1eoS1K40HLMl4zUyw5pnC05dEF3mzFLp5TPEEt4PFvsM
+[storage versions]:https://github.com/kubernetes/kubernetes/blob/db9545e69e630e5bc8c3c2dab14f9339dd3b871c/pkg/kubeapiserver/options/storage_versions.go#L98
+
+#### b. Using the apiserver version to trigger migration
+
+If the Kubernetes deprecation [policy][] is followed by every apiserver, and
+every cluster uses the default [storage version configs][], then we can use the
+apiserver version to drive the migration. More specifically,
+* kube-apiserver removes the `--storagve-version` flag, so that every deployment
+  of kube-apiserver uses the default storage versions. [#67678][]
+* aggregated apiservers follow semantic versioning, and expose their version via
+  API.
+* aggregated apiservers follow the same API deprecation [policy][].
+
+In both alternatives, the migration of custom resources is triggered by the
+difference between its [storage version][] and the [stored version][].
+
+[storage version configs]:https://github.com/kubernetes/kubernetes/blob/db9545e69e630e5bc8c3c2dab14f9339dd3b871c/pkg/kubeapiserver/options/storage_versions.go#L98
+[storage version]:https://github.com/kubernetes/kubernetes/blob/c1f7df2b0efe89fb9d26db2f2837bac11d57064e/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L66
+[stored version]:https://github.com/kubernetes/kubernetes/blob/c1f7df2b0efe89fb9d26db2f2837bac11d57064e/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types.go#L177
+[policy]:https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+[#67678]:https://github.com/kubernetes/kubernetes/pull/67678
+
+In both alternatives, HA apiservers present additional challenges. The migration
+must start after all HA apiservers have agreed on the storage versions or the
+apiserver versions. The agreement protocol is worth its own KEP.
+
+We will revisit this section in 2018 Q4. 
+
+### Risks and Mitigations
+
+The migration process does not change the objects, so it will not pollute
+existing data.
```

Comment:
"Before upgrading or downgrading the cluster, the cluster administrator must run
the `kube-migrator progress` to make sure the migration has completed".

Reply:
why not having this check automatically done inside kube-migrator? what if the admin forget to run "kube-migrator progress"?

Reply:
Perhaps I missed your point. "upgrading or downgrading the cluster" are done via cloud-specific tools. We cannot force those tools to run `kube-migrator progress`.

Reply:
We probably need the triggering mechanism to cancel an in-progress migration if the trigger changes (and then start a new one). It's e.g. safe for user to go forward and then back without waiting for the migration to complete.

But actually I don't know how to make this safe for a single object write (read on version X, but by the time there's a write it's version Y). You need something like a precondition on the api micro version for that to be 100% safe.

Reply:
> We probably need the triggering mechanism to cancel an in-progress migration if the trigger changes (and then start a new one). 

I think it's a general requirement to have a single migration job running for each resource type. I don't understand why it's a specific requirement of enabling fast roll back.

I didn't follow the second paragraph.

Reply:
Side note: to enable fast rollback, `migration.status` needs to reflect what is the currently stored in etcd, either the stored version, the hashed stored version, or the apiserver version (probably isn't safe in the future). I didn't put it in the proposal because we haven't reached consensus on the triggers.